### PR TITLE
RTOS_SwitchRead

### LINKED ISSUE #108: [RTOS] Read Switches Task
This RTOS task should read all of the input switches and set the car state accordingly. The state diagrams and some pseudocode can be found on the Google Drive (Electrical Team > Controls > New Design > Software > RTOS), but don't take either resource as absolute truth. This is the time to reason about this application code, so try to think about what exactly each module should be doing.

### PR DESCRIPTION
**Pull request by babusid**
Implemented the Read Switches task with the new task structure. Closes #108

### CODE REVIEW

File: Apps/Inc/CarState.h
Comment (JimothyGreene) [inline Apps/Inc/CarState.h:20]: Make this `FWD_SW` just to avoid confusion

File: Apps/Inc/CarState.h
Comment (JimothyGreene) [inline Apps/Inc/CarState.h:19]: This should be `HZD_SW`

File: Apps/Inc/StateTypes.h
Comment (JimothyGreene) [inline Apps/Inc/StateTypes.h:31]: We don't need this file anymore

File: Apps/Inc/Tasks.h
Comment (JimothyGreene) [inline Apps/Inc/Tasks.h:140]: This doesn't need to be global. It'll be passed to each task in `main()` as a `p_arg`

File: Drivers/Inc/Switches.h
Comment (JimothyGreene) [inline Drivers/Inc/Switches.h:13]: Same comments as above

File: Apps/Src/Tasks.c
Comment (JimothyGreene) [inline Apps/Src/Tasks.c:68]: Same as above

File: Apps/Src/ReadSwitches.c
Comment (JimothyGreene) [inline Apps/Src/ReadSwitches.c:20]: You're overwriting these variables immediately after writing to them

File: Apps/Src/ReadSwitches.c
Comment (JimothyGreene) [inline Apps/Src/ReadSwitches.c:10]: The logic here makes sense. A recommendation:

Instead of reading the switches grouped on how they're used, read all switches up front and update the `carState` struct. Then, you can go through each of the groupings and just read directly from the `carState` struct. That way, the `carState` values are always properly up to date.

**Review by JimothyGreene — CHANGES_REQUESTED** [review]
**Note:** I wouldn't merge this in until the ArrayConnection and MotorConnection tasks have been tested and merged. We want our `master` code to compile.

File: Apps/Src/ReadSwitches.c
Comment (babusid) [reply Apps/Src/ReadSwitches.c:10]: Don't the semaphores have to get signaled only if there is a difference between the prior car state and the new car state? If I overwrote the prior car state with the new values at the top, how would I know whether to signal or whether the switch had been on forever?

**Review by JimothyGreene — DISMISSED** [review]
Couple notes left

**Comment by babusid** [comment]
@Cam0Cow Do you recall why this hasn't been merged in just yet? Is it because we can't test it without the board?

**Comment by Cam0Cow** [comment]
> @Cam0Cow Do you recall why this hasn't been merged in just yet? Is it because we can't test it without the board?

Yes, it didn't make sense to merge it into master untested.

**Comment by babusid** [comment]
This Branch does not have the BSP CAN layer at present and as such is failing build checks. We need to merge in the CAN BSP layer into master, merge master into this branch, and test before we can merge this.

**Comment by babusid** [comment]
@Cam0Cow @chaseblock This is failing compilation when ` make stm32f413 TEST=SwitchRead` is run but not when ` make stm32f413 ` is run. Could either of you run the same thing and see if you catch any errors strictly in terms of compilation?

**Comment by ClarkPoon** [comment]
> @Cam0Cow @ClarkPoon This is failing compilation when ` make stm32f413 TEST=SwitchRead` is run but not when `make stm32f413` is run. Could either of you run the same thing and see if you catch any errors strictly in terms of compilation?

BSP_CAN is not implemented on this branch, so it can't find the BSP_CAN functions declared in BSP_CAN.h when it tries to link.

https://github.com/lhr-solar/Controls/blob/RTOS_SwitchRead/BSP/STM32F413/Src/BSP_CAN.C

**Comment by babusid** [comment]
> > @Cam0Cow @ClarkPoon This is failing compilation when ` make stm32f413 TEST=SwitchRead` is run but not when `make stm32f413` is run. Could either of you run the same thing and see if you catch any errors strictly in terms of compilation?
> 
> 
> 
> BSP_CAN is not implemented on this branch, so it can't find the BSP_CAN functions declared in BSP_CAN.h when it tries to link.
> 
> 
> 
> https://github.com/lhr-solar/Controls/blob/RTOS_SwitchRead/BSP/STM32F413/Src/BSP_CAN.C

Ah, these are the things you miss when you forget to sleep. I'll add dummy functions to the C file since this thread doesn't depend on CAN.

File: Apps/Src/ReadSwitches.c
Comment (ClarkPoon) [inline Apps/Src/ReadSwitches.c:67]: Is there a reason you are posting semaphores to control the contactors instead of just calling the contactor driver directly? It would get rid of the context switch overhead and reduce the number of tasks you need. I also think it would make the code more readable and easier to debug.

File: Apps/Src/ReadSwitches.c
Comment (ClarkPoon) [inline Apps/Src/ReadSwitches.c:183]: Why would you not just make a function to set the states of the different lights and call that? I don't see why you need to populate a global data structure and post a semaphore to make a task run to change the lights.

File: Apps/Src/ReadSwitches.c
Comment (ClarkPoon) [inline Apps/Src/ReadSwitches.c:219]: Why not use a function?

File: Drivers/Src/Switches.c
Comment (ClarkPoon) [inline Drivers/Src/Switches.c:4]: These names don't seem very intuitive to me

File: Drivers/Src/Switches.c
Comment (ClarkPoon) [inline Drivers/Src/Switches.c:29]:
```suggestion
    initTxBuf[2] = initRxBuf[1] | IODIRA_INIT;
```

File: Drivers/Src/Switches.c
Comment (ClarkPoon) [inline Drivers/Src/Switches.c:42]:
```suggestion
    initTxBuf[2] = initRxBuf[1] | IODIRB_INIT;
```

File: Tests/Test_SwitchRead.c
Comment (ClarkPoon) [inline Tests/Test_SwitchRead.c:11]: What is the point of declaring this `static` if you just pass a pointer to it into literally all of your tasks?

File: Apps/Src/ReadSwitches.c
Comment (babusid) [inline Apps/Src/ReadSwitches.c:84]: This task needs the CarState struct pointer

File: Apps/Src/ReadSwitches.c
Comment (babusid) [inline Apps/Src/ReadSwitches.c:112]: This task also needs the CarState struct pointer

File: Apps/Src/ReadSwitches.c
Comment (babusid) [inline Apps/Src/ReadSwitches.c:136]: Same as above

File: Apps/Src/ReadSwitches.c
Comment (IshDeshpa) [inline Apps/Src/ReadSwitches.c:13]: A bitmap that serves this same purpose exists in the BSP layer. Instead of using this to store the current switch states, use Switch_Read to access the bitmap.

File: Apps/Src/ReadSwitches.c
Comment (IshDeshpa) [inline Apps/Src/ReadSwitches.c:120]: Instead of updating a new CurrSwitches in this file, implementing a Switches_ReadBitmap function in the BSP layer might be a good idea. Or just use Switches_Read whenever necessary.

**Review by IshDeshpa — CHANGES_REQUESTED** [review]
- CurrSwitches is redundant because the switches bitmap exists on the BSP layer. 
- A Switches_ReadBitmap function on the BSP layer might be necessary.
- ArrayPrecharge is no longer a necessary state in the FSM.
- Use State struct correctly

File: Apps/Src/ReadSwitches.c
Comment (tiandahuang) [reply Apps/Src/ReadSwitches.c:46]: These functions and structs are used only for this task so they should probably stay within the .c

File: Apps/Src/ReadSwitches.c
Comment (tiandahuang) [reply Apps/Src/ReadSwitches.c:127]: CurrSwitches.HDLT gets its value from Read_Switches, so it'll be a valid State (ON/OFF).

File: Apps/Src/ReadSwitches.c
Comment (tiandahuang) [reply Apps/Src/ReadSwitches.c:135]: left/right blink gets its value from Read_Switches, so it'll be a valid State (ON/OFF)

File: Apps/Src/ReadSwitches.c
Comment (tiandahuang) [reply Apps/Src/ReadSwitches.c:50]: I think we still need to precharge the array

File: Apps/Src/ReadSwitches.c
Comment (tiandahuang) [reply Apps/Src/ReadSwitches.c:13]: This was written before we had that bitmap; I've updated it to use Switches_Read 😄👍

File: Apps/Src/ReadSwitches.c
Comment (IshDeshpa) [reply Apps/Src/ReadSwitches.c:50]: You are right, I was mistaken.

File: Apps/Src/ReadSwitches.c
Comment (babusid) [inline Apps/Src/ReadSwitches.c:102]: Reverse the order of these two, the point of having the array precharge is to reduce current spikes coming into the system. If we disconnect array Precharge before connecting the main contactor, we have a small period of 0 current, which sort of negates the point. While it's not a major concern, flipping them will be more in line with the designed behavior.
```suggestion
    Contactors_Set(ARRAY_CONTACTOR, ON);
    Contactors_Set(ARRAY_PRECHARGE, OFF);
```

File: Apps/Src/ReadSwitches.c
Comment (babusid) [inline Apps/Src/ReadSwitches.c:9]: This is just the interval we want to check the ignition switch for changes during array precharge right?

File: Apps/Src/ReadSwitches.c
Comment (babusid) [inline Apps/Src/ReadSwitches.c:76]: not a fan of this macro still

File: Apps/Src/ReadSwitches.c
Comment (babusid) [reply Apps/Src/ReadSwitches.c:94]: Upon re-reviewing, I see the global on line fifteen, but I have a strong feeling that down the line someone will copy this macro for usage elsewhere, and forget the global error variable that it needs. Let's just use the standard function.

File: Apps/Src/ReadSwitches.c
Comment (babusid) [inline Apps/Src/ReadSwitches.c:93]: If UpdateSwitches returns a zero, this function exits immediately. I believe the intended behavior of this is to exit the Array Precharge state and return to Electronics On. 
However, on Line 76, it delays a further ten milliseconds (based on the FSMs delay field of the precharge state). If during this small delay the switch gets turned to position three or back to position one (ie. someone turns it to zero and immediately back to 1 or three), line 77 of the FSM driver loop could potentially turn the array on before the precharge has actually finished.
I would suggest simply removing the 10-millisecond delay (or changing it to 1 ms) on Line 50 of the FSM table to solve this. While it is an edge case, I just want to bring it in line with the intended behavior. Let me know if I'm misinterpreting/missing something here.

**Review by babusid — COMMENTED** [review]
Also, can you make a task test file that spawns the ReadSwitches task so that we can test the task itself?

File: Apps/Src/ReadSwitches.c
Comment (babusid) [reply Apps/Src/ReadSwitches.c:167]: I agree with Ishan here, just for the sake of not casting when you don't have to, but it shouldn't be an issue.

Comment (tiandahuang) [reply Apps/Src/ReadSwitches.c:167]: Enums are int by default so there's not any actual casting going on. Also, I'm doing bitwise operations to yield left/right blink and it's probably best not to do bitwise operations on an enum.

File: Tests/Test_ReadSwitches.c
Comment (babusid) [inline Tests/Test_ReadSwitches.c:96]: Wrong TCB

File: Tests/Test_Switches.c
Comment (babusid) [inline Tests/Test_Switches.c:14]: FWD_SW isn't defined, I think it got replaced with FOR_SW

File: Tests/Test_ReadSwitches.c
Comment (babusid) [inline Tests/Test_ReadSwitches.c:79]: If we hit an OS error anywhere the program should ideally hang in a while loop or something so that it doesn't keep trying to execute and makes tracing the source more annoying.

**Review by babusid — COMMENTED** [review]
Not many changes needed, couple quick things in the testfiles, and making OS errors hang in the location that they happen. Can't merge this in before Minions is merged in however, and once Minion driver is merged into master, master needs to get merged into this branch.

File: Apps/Src/ReadSwitches.c
Comment (tiandahuang) [reply Apps/Src/ReadSwitches.c:8]: update switches gets called at every state, so if it loops at one state (like electronics on), then the delay there is the period that switches are updated

File: Apps/Src/Tasks.c
Comment (tiandahuang) [reply Apps/Src/Tasks.c:51]: Just took a look in master and I can't find it anywhere, should it be located in Tasks.c?

**Review by Cam0Cow — APPROVED** [review]

**Pull request merged**
